src/Deserialization/InstructionFactory.cs

```diff
@@ -0,0 +1,221 @@
+                LLVMOpcode.LLVMUserOp2 => throw new Exception("User op 2 should only be used internally in passes."),
+                LLVMOpcode.LLVMVAArg => throw new Exception("The front-end should lower va_arg."),
+                LLVMOpcode.LLVMExtractElement => new Unsupported(id, "extractelement"),
+                LLVMOpcode.LLVMInsertElement => new Unsupported(id, "insertelement"),
+                LLVMOpcode.LLVMShuffleVector => throw new Exception("The scalarizer pass should lower shufflevector."),
+                LLVMOpcode.LLVMExtractValue => new ExtractValue(
+                    id,
+                    operands,
+                    instruction.TypeOf.GetStoreSize(_targetData).ToBits(),
+                    GetConstantOffsets(instruction).ToArray()),
+                LLVMOpcode.LLVMInsertValue => new InsertValue(
+                    id,
+                    operands,
+                    GetConstantOffsets(instruction).ToArray()),
+                LLVMOpcode.LLVMFreeze => new Unsupported(id, "freeze"),
+                LLVMOpcode.LLVMFence => throw new Exception("The loweratomic pass should lower fence."),
+                LLVMOpcode.LLVMAtomicCmpXchg => throw new Exception("The loweratomic pass should lower cmpxchg."),
+                LLVMOpcode.LLVMAtomicRMW => throw new Exception("The loweratomic pass should lower atomicrmw."),
+                LLVMOpcode.LLVMResume => new Unsupported(id, "resume"),
+                LLVMOpcode.LLVMLandingPad => new Unsupported(id, "landingpad"),
+                LLVMOpcode.LLVMCleanupRet => new Unsupported(id, "cleanupret"),
+                LLVMOpcode.LLVMCatchRet => new Unsupported(id, "catchret"),
+                LLVMOpcode.LLVMCatchPad => new Unsupported(id, "catchpad"),
+                LLVMOpcode.LLVMCleanupPad => new Unsupported(id, "cleanuppad"),
+                LLVMOpcode.LLVMCatchSwitch => new Unsupported(id, "catchswitch"),
+                _ => throw new Exception("Instruction type is unknown.")
+            };
+        }
+
+        private Call CreateCall(InstructionId id, IOperand[] operands, LLVMValueRef instruction)
+        {
+            return new(
+                id,
+                operands,
+                instruction.TypeOf.GetStoreSize(_targetData).ToBits(),
+                GetAttributes(instruction, LLVMAttributeIndex.LLVMAttributeReturnIndex),
+                instruction.GetAttributeIndices()
+                    .Select(i => GetAttributes(instruction, i))
+                    .ToArray());
+        }
+
+        private IAttributes GetAttributes(LLVMValueRef instruction, LLVMAttributeIndex attributeIndex)
+        {
+            var attributes = instruction.GetCallSiteAttributes(attributeIndex);
+
+            var kind = _unsafeContext.GetEnumAttributeKind("signext");
+            var isSignExtended = attributes.Any(a => _unsafeContext.GetEnumAttributeKind(a) == kind);
+
+            return new Attributes(isSignExtended);
+        }
+
```

Comment:
I want full feedback on everything in this PR with the exception of the 2 functions below. They can clearly be golfed but I want to improve their actual representations which will affect this quite a bit. I want to make those improvements in a separate PR so this change is kept inside the Deserialization space and is a fairly direct translation from the C++ equivalent.

Reply:
Ignored.